pkg/kubelet/images/image_manager.go

```diff
@@ -177,3 +245,57 @@ func applyDefaultImageTag(image string) (string, error) {
 	}
 	return image, nil
 }
+
+// isEnsuredBySecret - returns true if the secret for an auth used to pull an
+// image has already been authenticated through a successful pull request
+// and the same auth exists for this podSandbox/image/
+func (m *imageManager) isEnsuredBySecret(imageRef string, image kubecontainer.ImageSpec, pullSecrets []v1.Secret) (pulledBySecret, ensuredBySecret bool) {
+	if imageRef == "" {
+		return
+	}
+	if ensureSecretPulledImage[imageRef] != nil {
+		pulledBySecret = true
+	}
+
+	img := image.Image
+	repoToPull, _, _, err := parsers.ParseImageName(img)
+	if err != nil {
+		return
+	}
+
+	if m.keyring == nil {
+		return
+	}
+	keyring, err := credentialprovidersecrets.MakeDockerKeyring(pullSecrets, *m.keyring)
+	if err != nil {
+		return
+	}
+
+	creds, withCredentials := keyring.Lookup(repoToPull)
+	if !withCredentials {
+		return
+	}
+
+	for _, currentCreds := range creds {
+		auth := &runtimeapi.AuthConfig{
+			Username:      currentCreds.Username,
+			Password:      currentCreds.Password,
+			Auth:          currentCreds.Auth,
+			ServerAddress: currentCreds.ServerAddress,
+			IdentityToken: currentCreds.IdentityToken,
+			RegistryToken: currentCreds.RegistryToken,
+		}
+
+		hash := kubecontainer.HashAuth(auth)
+		if hash != "" {
```

Comment:
When can this be empty?

Reply:
should never happen

Reply:
would require a failure to marshal auth.. fallback would be to fall back to prior (slower) 
behavior

Reply:
going to sleep on this..